app/apps/handlers.js

```diff
@@ -20,12 +20,25 @@ exports.create = (req, res, next) => {
     userapps: [],
   });
   let new_app;
+  let new_bucket;
+
+  if(req.body['new-app-datasource'] === 'create') {
+    new Bucket({
+      name: req.body['new-datasource-name']
+    }).create()
+      .then((created_bucket) => {
+        new_bucket = created_bucket;
+      });
+  }
```

Comment:
This might not work - the `Bucket.create` promise here might not resolve before the `app.create` promise below.

You could nest the promises, eg:
```javascript
app.create()
  .then((created_app) => {
    ...
    if (req.body['new-app-datasource'] === 'create') {
      new Bucket({ name: req.body['new-datasource-name'] }).create()
        .then((created_bucket) => {
          created_app.grant_bucket_access(created_bucket.id, 'readonly');
        });
    }
  });
```

Or you could run them in parallel, eg:
```javascript
const create_bucket = new Promise((resolve, reject) => {
  if (req.body['new-app-datasource'] === 'create') {
    resolve(new Bucket({...}).create());
  } else {
    resolve(null);
  }
});
Promise.all([app.create(), create_bucket])
  .then([created_app, created_bucket]) => {
    if (created_bucket) {
      created_app.grant_bucket_access(created_bucket.id, 'readonly');
    }
    ...
  });
```

Reply:
👍  Thanks, will give the latter a go.